.clang-tidy

```diff
@@ -0,0 +1,315 @@
+---
+# This file defines what clang-tidy checks and parameter. It is experimental, and may let 
```

Comment:
@envoyproxy/maintainers I am open to any suggestion to this file. With this `.clang-tidy` it still issue too many warnings. A result from a couple days ago:
https://gist.github.com/lizan/115f05ace9ac7dbae4c042f6fbdb3596

Reply:
Hmm. Don't we need to fail if clang-tidy fails? So don't we need to be warning clean? Or we only fail on errors? In general it seems like we should be warning clean also.

Reply:
No I don't think we should mark clang-tidy required, at least in a meanwhile. We can make some of checks error with `-warnings-as-errors` later to fail on them.

Reply:
Then what's the point of using a CI slot for this? No one will look at it.

Reply:
Just like mac CI that is not required (on GitHub) but we still check? At this point, there are too many warnings to fix at once so I am hesitating to make them error.

cc @jsedgwick @rlazarus @PiotrSikora on #921, any suggestion to enable some checks as `WarningsAsErrors`?

Reply:
How does Google deal with this internally? Since Google fixes bugs raised by clang-tidy presumably some internal CI is catching them? Can we just start with Google's config? cc @htuch @alyssawilk @PiotrSikora

Reply:
Internally clang-tidy is largely advisory I believe, but most reviewers will take its suggestions as actionable. For Envoy, only some importers (we have a weekly rotation) actually pay attention as it's largely mechanical.

I think we have a noise problem with a purely advisory clang-tidy, if the report is as noisy as in https://gist.github.com/lizan/115f05ace9ac7dbae4c042f6fbdb3596, reviewers will never both to look at it. Inside Google, our review system highlights issues only in the delta diff.

Maybe we can turn off all warnings, make clang-tidy compulsory and only turns on the ones that actually are low on the false positive side?

Reply:
Yes running clang-tidy on diff (especially for PRs) is definitely the plan, considering how clang-tidy is heavy (it takes an hour to run on whole source code). There is [a tool](https://github.com/llvm-mirror/clang-tools-extra/blob/release_70/clang-tidy/tool/clang-tidy-diff.py) to run clang-tidy against diff.

If we do have a better .clang-tidy to start with, I'm happy to start with that rather than this.

Reply:
+1, IMO we need to make this actionable or we shouldn't bother running it in CI.